### PR TITLE
feat: added the ability to create cog

### PR DESCRIPTION
**Pull request by Ben-Cheng**

### CODE REVIEW

File: topo_processor/subprocess/create_cog.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        --user $CUID:$CGID \
+        --volume $PWD:$PWD \
+        --workdir $PWD \
+        --rm \
+        osgeo/gdal:ubuntu-small-latest $@
+}
+
+
+baseFileName=$(basename $INPUT_FILE)
+docker_gdal gdal_translate $INPUT_FILE -of COG -co COMPRESS=deflate $TRANSLATE_ARGS $OUTPUT_FOLDER/${baseFileName%.tif}.deflate.cog.tiff
```
Comment (Ben-Cheng) [inline topo_processor/subprocess/create_cog.sh:21]: shell scripts don't run on Windows.
All of this should live inside of Python

File: topo_processor/cog/create_cog.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+from linz_logger import get_log
+
+from topo_processor.util.time import time_in_ms
+
+
+def docker_gdal():
+    return f"docker run \
+        --user {os.geteuid()}:{os.getegid()} \
+        --volume {os.getcwd()}:{os.getcwd()} \
```
Comment (blacha) [inline topo_processor/cog/create_cog.py:12]: this path needs to include both the input and output directories

File: topo_processor/cog/create_cog.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        --workdir {os.getcwd()} \
+        --rm \
+        osgeo/gdal:ubuntu-small-latest $@"
+
+
+async def create_cog(input_tiff_path, output_dir):
+    start_time = time_in_ms()
+    translate_args = "-co NUM_THREADS=ALL_CPUS -co PREDICTOR=2"
```
Comment (blacha) [inline topo_processor/cog/create_cog.py:20]: this should be a const somewhere

File: topo_processor/cog/create_cog.py (hunk 2 above)
Comment (blacha) [reply topo_processor/cog/create_cog.py:14]: This isnt going to work, you need to find the best paths for input and output, CWD is not needed.

```python

volumes = []
if input.startsWith(output):
  volumes.push(output)
else if output.startsWith(input):
  volumes.push(input)
else if output == input:
 volumes.push(input)
else:
  volumes.push(input, output)
```

File: topo_processor/cog/create_cog.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+TRANSLATE_ARGS = "-co NUM_THREADS=ALL_CPUS -co PREDICTOR=2"
+
+
+def docker_gdal(output_dir: str):
+    return f"docker run \
+        --user {os.geteuid()}:{os.getegid()} \
+        --volume {os.getcwd()}:{os.getcwd()} \
+        --volume {output_dir}:{output_dir} \
+        --workdir {os.getcwd()} \
```
Comment (blacha) [inline topo_processor/cog/create_cog.py:16]: can remove the workdir too

**Comment by MDavidson17** [comment]
This code runs and will produce a COG, my current issue though is where to put the call to run `ExecutionDocker.to_docker()`.

Initially, it was in line 43 of execution.py however this put the gdal args ahead of the docker args so the command looked like this:
```
['docker', 'gdal_translate', '/home/mdavidson/dev/topo-processor/test_data/tiffs/CROWN_399_E_49.tiff', '-of', 'COG', '-co', 'COMPRESS=lzw', '-co', 'NUM_THREADS=ALL_CPUS', '-co', 'PREDICTOR=2', '/home/mdavidson/dev/topo-processor/CROWN_399_E_49.tiff.LZW.cog.tiff', 'run', '--user', '1000:1000', '-v', '/home/mdavidson/dev/topo-processor:/home/mdavidson/dev/topo-processor', '--rm', 'osgeo/gdal:ubuntu-small-latest']
```

I then moved it to line 17 of command.py, but this means the volumes weren't added:
```
['docker', 'run', '--user', '1000:1000', '--rm', 'osgeo/gdal:ubuntu-small-latest', 'gdal_translate', '/home/mdavidson/dev/topo-processor/test_data/tiffs/CROWN_399_E_49.tiff', '-of', 'COG', '-co', 'COMPRESS=lzw', '-co', 'NUM_THREADS=ALL_CPUS', '-co', 'PREDICTOR=2', '/home/mdavidson/dev/topo-processor/CROWN_399_E_49.tiff.LZW.cog.tiff']
```

Currently, it is in line 12 of create_cog.py, it works here however I don't think this is the best place for it. 

@blacha It would be great if you could take another look and let me know if we've made any mistakes with the structure or where you think its best to add the `ExecutionDocker.to_docker()` call.

File: topo_processor/cog/create_cog.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+from topo_processor.cog.command import Command
+from topo_processor.cog.execution import ExecutionDocker
+
+
+async def create_cog(input_file: str, output_dir: str, volumes: List[str], compression_method: str):
+    gdal = Command("docker", {"container": "osgeo/gdal", "tag": "ubuntu-small-latest"})
+    for volume in volumes:
+        gdal.mount(volume)
+    gdal = ExecutionDocker.to_docker(gdal)
```
Comment (blacha) [inline topo_processor/cog/create_cog.py:12]: You should never run this function outside of the `Execution` classes

File: topo_processor/cog/execution.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        docker.arg("--user", f"{os.geteuid()}:{os.getegid()}")
+        for volume in cmd.volumes:
+            docker.arg("-v", f"{volume}:{volume}")
+        docker.arg("--rm")
+        docker.arg(f"{docker.container}:{docker.container_tag}")
+        return docker
+
+    @staticmethod
+    async def run(cmd: "Command"):
+        return await ExecutionLocal.run(cmd)
```
Comment (blacha) [inline topo_processor/cog/execution.py:46]: this is where the `to_docker` needs to be run

File: topo_processor/cog/create_cog.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import os
+from typing import List
+
+from topo_processor.cog.command import Command
+from topo_processor.cog.execution import ExecutionDocker
+
+
+async def create_cog(input_file: str, output_dir: str, volumes: List[str], compression_method: str):
+    gdal = Command("docker", {"container": "osgeo/gdal", "tag": "ubuntu-small-latest"})
```
Comment (blacha) [inline topo_processor/cog/create_cog.py:9]: this command should be `gdal_translate`

File: topo_processor/cog/create_cog.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import os
+from typing import List
+
+from topo_processor.cog.command import Command
+from topo_processor.cog.execution import ExecutionDocker
+
+
+async def create_cog(input_file: str, output_dir: str, volumes: List[str], compression_method: str):
```
Comment (blacha) [inline topo_processor/cog/create_cog.py:8]: the `volumes` are `input_file` and `output_dir` and should not need another list

`compression_method` is not actually used in the cog creation

File: topo_processor/cog/create_cog_test.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    output_dir = os.getcwd()
+    compression_method = "LZW"
+    await create_cog(input_file, output_dir, compression_method)
+
+
+@pytest.mark.asyncio
+async def test_command_local():
+    cmd = Command("ls")
+    cmd.append_arg("-l")
+    await cmd.run()
```
Comment (blacha) [inline topo_processor/cog/create_cog_test.py:30]: What is this actually testing? need some sort of assertion about what is returned, exit code, outputs etc..

File: topo_processor/cog/command.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    def append_arg(self, *args: str) -> "Command":
+        for argument in args:
+            self.arguments.append(argument)
+        return self
+
+    def mount(self, *args: str) -> "Command":
+        """Mount a folder, useful only if the command is run inside of docker"""
+        for volume in args:
+            self.volumes.append(volume)
+        return os.sendfile
```
Comment (blacha) [inline topo_processor/cog/command.py:35]: this looks like a typo

File: topo_processor/cog/command.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        return [self.command] + self.arguments
+
+    def to_docker(self) -> "Command":
+        if not self.container:
+            raise Exception(f"No container found for command {self.command}")
+        docker = Command("docker")
+        docker.append_arg("run")
+        docker.append_arg("--user", f"{os.geteuid()}:{os.getegid()}")
+        for volume in self.volumes:
+            docker.mount(volume)
```
Comment (blacha) [inline topo_processor/cog/command.py:47]: dont need this mount

File: topo_processor/cog/execution.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        proc = await asyncio.create_subprocess_exec(
+            *cmd.to_full_command(), stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+        )
+        stdout, stderr = await proc.communicate()
+        if proc.returncode != 0:
+            raise Exception(stderr.decode())
+        get_log().debug(
+            "Ran command", command=cmd.to_full_command(), stdout=stdout.decode(), duration=time_in_ms() - start_time
+        )
+        return proc.returncode
```
Comment (blacha) [inline topo_processor/cog/execution.py:27]: you should return `{returnCode, stdout, stderr}` as a object rather than using `capsys`

**Review by blacha — APPROVED** [review]

**Pull request merged**